python/mlx/nn/layers/normalization.py

```diff
@@ -178,3 +180,139 @@ def __call__(self, x):
         )
         x = group_norm(x)
         return (self.weight * x + self.bias) if "weight" in self else x
+
+
+class BatchNorm(Module):
+    r"""Applies Batch Normalization over a 2D or 3D input.
+
+    Computes
+
+    .. math::
+
+        y = \frac{x - E[x]}{\sqrt{Var[x]} + \epsilon} \gamma + \beta,
+
+    where :math:`\gamma` and :math:`\beta` are learned per feature dimension
+    parameters initialized at 1 and 0 respectively.
+
+    [1]: https://arxiv.org/abs/1502.03167
+
+    The input tensor shape is specified as (N, C) or (N, C, L), representing the batch size (N), the number of features or channels (C), and optionally, the sequence length (L). The output tensor maintains the same shape as the input, adhering to (N, C) or (N, C, L).
+    For three-dimensional tensors, the shape is denoted as (N, C, H, W), where N signifies the batch size, C represents the number of channels, H corresponds to the height, and W denotes the width.
```

Comment:
This looks great!

I think one change can make it more consistent and a lot simpler:

For our convolutions (and in general) we follow the convention that the channels are last. So inputs to convolutions are `NLC` or `NHWC`. We should change two thigns:

1. Batch norm should also follow that convention
2. Since it is following that convention it should easily broadcast with the inputs and you can remove the whole check_and_expand_dims machinery and just let broadcasting manage it (it's super cheap to expand dims at runtime so from a perf perspective it should be trivial!)

Reply:
done, I've updated the batch norm implementation and tests to handle inputs of shape, NLC, NWHC!